### DISCUSSION TITLE
Photopea Integration - New Feature?

### DISCUSSION THREAD

**Discussion by InfraRd** [Mikubill/sd-webui-controlnet#2375] [Q&A] [answered]
I noticed that there is "Edit" button in preprocessor preview which pop ups a classic Photopea window. Does it use API to connect Photopea and send and receive the images to edit? Or does it use local hardware? (According to Terms of Use, it says files are not sent to anywhere but is this possible?)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/144592496/5201c05f-a995-4d56-aa50-fc1d961a85ad)

**Comment by huchenlei** [comment]
Photopea is a pure JS implemented image editor. The privacy policy means it does not send the image to a remote server and potentially persist it.

**Reply by huchenlei** [reply, marked as answer]
You can see this PR for details of implementation and how to use it: https://github.com/Mikubill/sd-webui-controlnet/pull/2346

The way ControlNet code communicate with Photopea is via frame message, which is local to your device.

**Reply by InfraRd** [reply]
Oh, i see now. Thanks for the reply and clarification!
